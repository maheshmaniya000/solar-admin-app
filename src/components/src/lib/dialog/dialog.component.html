<div mat-dialog-title>
  <button
    *ngIf="closeIconButtonEnabled"
    [mat-dialog-close]="undefined"
    class="close-button"
    mat-icon-button
    data-test="close-button"
  >
    <mat-icon>close</mat-icon>
  </button>
  <h3>{{ titleTranslationKey | transloco }}</h3>
</div>
<mat-divider *ngIf="dividedContent" class="top-content-divider"></mat-divider>
<mat-dialog-content [ngClass]="{ divided: dividedContent }">
  <ng-content></ng-content>
</mat-dialog-content>
<mat-divider
  *ngIf="dividedContent"
  class="bottom-content-divider"
></mat-divider>
<ng-content select="mat-dialog-actions"></ng-content>
<mat-dialog-actions *ngIf="!actionsDirective" align="end">
  <ng-container *ngFor="let action of actions | async">
    <button
      *ngIf="!action.dismissing; else dismissingButton"
      [disabled]="action.disabled"
      [mat-dialog-close]="action.payload"
      cdkFocusInitial
      color="accent"
      mat-flat-button
      [attr.data-test]="action.dataTest"
    >
      {{ action.textTranslationKey | transloco }}
    </button>
    <ng-template #dismissingButton>
      <button
        [disabled]="action.disabled"
        [mat-dialog-close]="action.payload"
        color="accent"
        mat-stroked-button
        [attr.data-test]="action.dataTest"
      >
        {{ action.textTranslationKey | transloco }}
      </button>
    </ng-template>
  </ng-container>
</mat-dialog-actions>
<ng-container *ngIf="footer">
  <mat-divider class="footer-divider"></mat-divider>
  <ng-content select="sg-dialog-footer"></ng-content>
</ng-container>
