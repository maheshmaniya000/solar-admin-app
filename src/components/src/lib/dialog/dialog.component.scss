$vertical-content-spacing: 12px;

.mat-dialog-title {
  display: flex;
  align-items: center;

  h3 {
    margin-bottom: 0;
  }
}

.mat-dialog-content {
  padding-top: $vertical-content-spacing;
  padding-bottom: $vertical-content-spacing;

  ::ng-deep .mat-divider-vertical {
    margin-top: -$vertical-content-spacing;
    margin-bottom: -$vertical-content-spacing;
  }
}

:host ::ng-deep .sg-dialog-footer {
  display: flex;
  align-items: center;
  min-height: 48px;
}
