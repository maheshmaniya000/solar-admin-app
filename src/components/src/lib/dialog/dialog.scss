@use 'variables';

$medium-dialog-vertical-spacing: 20px;
$medium-dialog-horizontal-spacing: 32px;
$medium-dialog-actions-min-height: 56px;

$large-dialog-vertical-spacing: 28px;
$large-dialog-horizontal-spacing: 48px;
$large-dialog-actions-min-height: 68px;

$mat-dialog-actions-vertical-padding: 8px;

$medium-button-height: map-get(variables.$sg-button-height, medium);
$large-button-height: map-get(variables.$sg-button-height, large);

@mixin _divider-horizontal-spacing($value) {
  margin-right: -$value;
  margin-left: -$value;
}

@mixin _dialog(
  $vertical-spacing,
  $horizontal-spacing,
  $actions-min-height,
  $button-height,
  $title-margin-bottom: null
) {
  padding: $vertical-spacing $horizontal-spacing;

  @if $title-margin-bottom {
    .mat-dialog-title {
      margin-bottom: $title-margin-bottom;
    }
  }

  .mat-dialog-content {
    margin-right: -$horizontal-spacing;
    margin-left: -$horizontal-spacing;
    padding-right: $horizontal-spacing;
    padding-left: $horizontal-spacing;
  }

  .top-content-divider,
  .bottom-content-divider {
    @include _divider-horizontal-spacing($horizontal-spacing);
  }

  .mat-dialog-actions {
    min-height: $actions-min-height;
    margin-bottom: -($vertical-spacing - $mat-dialog-actions-vertical-padding);
  }

  .sg-dialog-footer {
    margin-bottom: -$vertical-spacing;
  }

  .footer-divider {
    margin-top: $vertical-spacing;
    @include _divider-horizontal-spacing($horizontal-spacing);
  }
}

.sg-dialog-container {
  .mat-dialog-container {
    .close-button {
      margin-right: 6px;
      margin-left: -6px;
    }

    @include _dialog(
      $medium-dialog-vertical-spacing,
      $medium-dialog-horizontal-spacing,
      $medium-dialog-actions-min-height,
      $medium-button-height
    );
  }

  &.large {
    .mat-dialog-container {
      @include _dialog(
        $large-dialog-vertical-spacing,
        $large-dialog-horizontal-spacing,
        $large-dialog-actions-min-height,
        $large-button-height,
        $large-dialog-vertical-spacing
      );
    }

    .mat-dialog-actions button {
      height: $large-button-height;
    }

    &.featured {
      .mat-dialog-title,
      .mat-dialog-content,
      .mat-dialog-actions,
      .sg-dialog-footer {
        justify-content: center;
      }

      .mat-dialog-title {
        position: relative;
        margin-bottom: 0;
      }

      .close-button {
        position: absolute;
        top: -$large-dialog-vertical-spacing + 6px;
        left: -$large-dialog-horizontal-spacing + 12px;
      }

      .mat-dialog-content {
        text-align: center;
      }
    }
  }
}
