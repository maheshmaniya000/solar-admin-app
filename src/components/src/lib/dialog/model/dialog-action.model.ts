import { Observable } from 'rxjs';

export interface DialogAction<T = any> {
  dismissing: boolean;
  textTranslationKey: string;
  payload: T;
  disabled?: boolean;
  dataTest: string;
}

// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays
type TypesToDialogActions<Types extends any[]> = { [K in keyof Types]: DialogAction<Types[K]> };

export type DialogActions<PayloadTypes extends any[] = any[]> =
  | TypesToDialogActions<PayloadTypes>
  | Observable<TypesToDialogActions<PayloadTypes>>;
