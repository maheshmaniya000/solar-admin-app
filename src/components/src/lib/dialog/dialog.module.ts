import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import {
  MatDialogActions,
  MatDialogClose,
  MatDialogModule,
  MAT_DIALOG_DEFAULT_OPTIONS
} from '@angular/material/dialog';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule } from '@angular/material/icon';
import { TranslocoModule } from '@ngneat/transloco';

import { DialogComponent } from './dialog.component';
import { DialogConstants } from './dialog.constants';
import { DialogFooterDirective } from './footer-content/dialog-footer.directive';
import { StandardDialogComponent } from './standard-dialog/standard-dialog.component';
import { SuccessDialogContentComponent } from './success-dialog/success-dialog-content.component';

@NgModule({
  declarations: [
    StandardDialogComponent,
    DialogComponent,
    DialogFooterDirective,
    SuccessDialogContentComponent
  ],
  imports: [
    CommonModule,
    MatButtonModule,
    MatDialogModule,
    MatDividerModule,
    MatIconModule,
    TranslocoModule
  ],
  exports: [
    MatDialogActions,
    StandardDialogComponent,
    DialogComponent,
    DialogFooterDirective,
    MatDialogClose
  ],
  providers: [
    {
      provide: MAT_DIALOG_DEFAULT_OPTIONS,
      useValue: DialogConstants.defaultConfigs.medium
    }
  ]
})
export class DialogModule {}
