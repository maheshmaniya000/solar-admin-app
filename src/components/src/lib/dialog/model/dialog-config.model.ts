import { MatDialogConfig } from '@angular/material/dialog';

export type DialogConfig<D = any> = Pick<MatDialogConfig<D>, 'data' | 'disableClose'>;
