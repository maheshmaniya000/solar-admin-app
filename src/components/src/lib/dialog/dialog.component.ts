import {
  AfterContentChecked,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChild,
  Input,
  OnChanges,
  OnInit
} from '@angular/core';
import { MatDialogActions } from '@angular/material/dialog';
import { isNil } from 'lodash-es';
import { isObservable,  of } from 'rxjs';

import { DialogConstants } from './dialog.constants';
import { DialogFooterDirective } from './footer-content/dialog-footer.directive';
import { DialogActions } from './model/dialog-action.model';

@Component({
  selector: 'sg-dialog',
  templateUrl: './dialog.component.html',
  styleUrls: ['./dialog.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class DialogComponent implements OnInit, OnChanges, AfterContentChecked {
  @Input() titleTranslationKey: string;
  @Input() dividedContent = false;
  @Input() closeIconButtonEnabled = false;
  @Input() actions: DialogActions;
  @ContentChild(MatDialogActions) actionsDirective: MatDialogActions;
  @ContentChild(DialogFooterDirective)
  footer: DialogFooterDirective;

  constructor(private readonly changeDetectorRef: ChangeDetectorRef) {}

  ngOnChanges(): void {
    if (isNil(this.actions)) {
      this.actions = of([
        DialogConstants.actions.cancel,
        DialogConstants.actions.ok
      ]);
    }
    this.actions = isObservable(this.actions) ? this.actions : of(this.actions);
  }

  ngOnInit(): void {
    this.ngOnChanges();
  }

  ngAfterContentChecked(): void {
    this.changeDetectorRef.markForCheck();
  }
}
