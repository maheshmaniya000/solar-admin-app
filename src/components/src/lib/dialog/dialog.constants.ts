import { MatDialogConfig } from '@angular/material/dialog';

import { DialogAction } from './model/dialog-action.model';
import { DialogKind } from './model/dialog-kind.model';

type DialogActionKind = 'cancel' | 'ok' | 'delete' | 'change';

export class DialogConstants {
  static readonly confirmAction: Omit<DialogAction<true>, 'textTranslationKey' | 'dataTest'> = {
    dismissing: false,
    payload: true
  };

  static readonly actions: Record<DialogActionKind, DialogAction<boolean>> = {
    cancel: {
      dismissing: true,
      textTranslationKey: 'common.action.cancel',
      payload: false,
      dataTest: 'cancel-action-button'
    },
    ok: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.ok',
      dataTest: 'ok-action-button'
    },
    delete: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.delete',
      dataTest: 'delete-action-button'
    },
    change: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.change',
      dataTest: 'change-action-button'
    }
  };

  private static readonly sgDialogContainerClass = 'sg-dialog-container';
  private static readonly sgLargeDialogContainerClass = 'large';
  private static readonly sgFeaturedDialogContainerClass = 'featured';
  private static readonly defaultMediumDialogWidth = '608px';
  private static readonly defaultLargeDialogWidth = '640px';

  static readonly defaultConfigs: Record<DialogKind, MatDialogConfig> = {
    medium: {
      ...new MatDialogConfig(),
      panelClass: DialogConstants.sgDialogContainerClass,
      width: DialogConstants.defaultMediumDialogWidth
    },
    large: {
      ...new MatDialogConfig(),
      panelClass: [
        DialogConstants.sgDialogContainerClass,
        DialogConstants.sgLargeDialogContainerClass
      ],
      width: DialogConstants.defaultLargeDialogWidth
    },
    featured: {
      ...new MatDialogConfig(),
      panelClass: [
        DialogConstants.sgDialogContainerClass,
        DialogConstants.sgLargeDialogContainerClass,
        DialogConstants.sgFeaturedDialogContainerClass
      ],
      width: DialogConstants.defaultLargeDialogWidth
    }
  };

  static createStandardDialogDefaultActions<T>(
    okActionPayload?: T,
    okActionDisabled: boolean = false
  ): [DialogAction<undefined>, DialogAction<T>] {
    return [
      { ...DialogConstants.actions.cancel, payload: undefined },
      {
        ...DialogConstants.actions.ok,
        payload: okActionPayload,
        disabled: okActionDisabled
      }
    ];
  }
}
