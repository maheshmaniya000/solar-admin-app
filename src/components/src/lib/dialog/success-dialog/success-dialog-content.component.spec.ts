import { MAT_DIALOG_DATA } from '@angular/material/dialog';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { TranslocoMocks } from 'ng-shared/utils/transloco.mocks';

import { DialogConstants } from '../dialog.constants';
import { SuccessDialogData } from './models/success-dialog-data.model';
import { SuccessDialogContentComponent } from './success-dialog-content.component';

describe('SuccessDialogContentComponent', () => {
  describe('component', () => {
    let spectator: Spectator<SuccessDialogContentComponent>;
    const createComponent = createComponentFactory({
      component: SuccessDialogContentComponent,
      imports: [TranslocoMocks.createTestingModule()],
      providers: [
        {
          provide: MAT_DIALOG_DATA,
          useValue: {
            imageUrl: 'some-image.jpg',
            headerTranslationKey: 'some.successHeader',
            actions: [DialogConstants.actions.ok]
          } as SuccessDialogData
        }
      ],
      shallow: true
    });

    beforeEach(() => {
      spectator = createComponent();
    });

    it(`should match snapshot without the provided success text`, () => {
      expect(spectator.fixture).toMatchSnapshot();
    });

    it(`should match snapshot with the provided success text`, () => {
      spectator.component.data.textTranslationKey = 'some.successText';
      spectator.detectComponentChanges();
      expect(spectator.fixture).toMatchSnapshot();
    });

    it(`should match snapshot with the provided image alt text`, () => {
      spectator.component.data.imageAltTextTranslationKey = 'some.successAltText';
      spectator.detectComponentChanges();
      expect(spectator.fixture).toMatchSnapshot();
    });
  });
});
