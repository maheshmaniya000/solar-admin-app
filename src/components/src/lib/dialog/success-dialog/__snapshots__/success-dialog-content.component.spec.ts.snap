// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuccessDialogContentComponent component should match snapshot with the provided image alt text 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  data={[Function Object]}
>
  <div
    class="image-container"
  >
    <img
      alt="en.some.successAltText"
      src="some-image.jpg"
    />
  </div><div
    class="mat-display-2"
  >
    en.some.successHeader
  </div><p>
    en.some.successText
  </p>
</ng-component>
`;

exports[`SuccessDialogContentComponent component should match snapshot with the provided success text 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  data={[Function Object]}
>
  <div
    class="image-container"
  >
    <img
      alt="en.some.successHeader"
      src="some-image.jpg"
    />
  </div><div
    class="mat-display-2"
  >
    en.some.successHeader
  </div><p>
    en.some.successText
  </p>
</ng-component>
`;

exports[`SuccessDialogContentComponent component should match snapshot without the provided success text 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  data={[Function Object]}
>
  <div
    class="image-container"
  >
    <img
      alt="en.some.successHeader"
      src="some-image.jpg"
    />
  </div><div
    class="mat-display-2"
  >
    en.some.successHeader
  </div>
</ng-component>
`;
